pkg/controller/elasticsearch/certificates/reconcile.go

```diff
@@ -54,58 +54,35 @@ func Reconcile(
 		results.WithError(err)
 	}
 
-	labels := label.NewLabels(k8s.ExtractNamespacedName(&es))
-
-	httpCA, err := certificates.ReconcileCAForOwner(
-		driver.K8sClient(),
-		esv1.ESNamer,
-		&es,
-		labels,
-		certificates.HTTPCAType,
-		caRotation,
-	)
-	if err != nil {
-		return nil, results.WithError(err)
-	}
-
-	// make sure to requeue before the CA cert expires
-	results.WithResult(reconcile.Result{
-		RequeueAfter: certificates.ShouldRotateIn(time.Now(), httpCA.Cert.NotAfter, caRotation.RotateBefore),
-	})
-
-	// discover and maybe reconcile for the http certificates to use
-	httpCertificates, err := http.ReconcileHTTPCertificates(
-		driver,
-		&es,
-		esv1.ESNamer,
-		httpCA,
-		es.Spec.HTTP.TLS,
-		labels,
-		services,
-		certRotation,
-	)
-	if err != nil {
-		return nil, results.WithError(err)
-	}
-
-	primaryCert, err := certificates.GetPrimaryCertificate(httpCertificates.CertPem())
-	if err != nil {
-		return nil, results.WithError(err)
-	}
-	results.WithResult(reconcile.Result{
-		RequeueAfter: certificates.ShouldRotateIn(time.Now(), primaryCert.NotAfter, certRotation.RotateBefore),
-	})
-
-	// reconcile http public certs secret:
-	if err := http.ReconcileHTTPCertsPublicSecret(driver.K8sClient(), &es, esv1.ESNamer, httpCertificates); err != nil {
-		return nil, results.WithError(err)
+	// label certificates secrets with the cluster name
+	certsLabels := label.NewLabels(k8s.ExtractNamespacedName(&es))
+
+	// reconcile HTTP CA and cert
+	var httpCerts *certificates.CertificatesSecret
+	httpCerts, results = certificates.Reconciler{
+		K8sClient:      driver.K8sClient(),
+		DynamicWatches: driver.DynamicWatches(),
+		Object:         &es,
+		TLSOptions:     es.Spec.HTTP.TLS,
+		Namer:          esv1.ESNamer,
+		Labels:         certsLabels,
+		Services:       services,
+		CACertRotation: caRotation,
+		CertRotation:   certRotation,
+		// ES is able to hot-reload TLS certificates: let's keep secrets around even though TLS is disabled.
```

Comment:
That's a good question, I'm scratching my head a bit over it.

For other resources (APMServer, Kibana, Enterprise Search), we rotate Pods automatically when there's a change in TLS certs (enabled, disabled, or different content - we hash it). So we don't care much about keeping secrets around.

For ES, we rotate Pods on config change (TLS enabled/disabled), but not on certs content change. Pods rotation can take a long time (hours for very big clusters) since we wait for the cluster to be green in-between each Pod.
If we remove certs secrets, their content remain in the Pods that had them mounted, until those Pods get deleted. However I think if the Secrets then get recreated, their content does not propagate to the Pods that had the old secret mounted.
So I think the following may happen:
- user disables TLS
- we remove all certs (their old content stay there in the Pods)
- we start a 2 hours long rolling upgrade where we rotate Pods which do not have secrets mounted anymore
- halfway through users change their mind and re-enable TLS
- at this point half Pods don't have TLS secrets mounted (the ones that have been rotated), the other half have what remains of the old deleted secrets (the ones that have not been rotated)
- the first half will get recreated again (rolling upgrade) to have the certs secrets mounted
- the second half will not get recreated since their configuration is ok (TLS enabled), however their certs content will never be updated with the new ones

Reply:
Hmm actually after doing some test what I said above is wrong: secret that are recreated still seem to propagate to Pods that have them mounted.
Will do more tests.

Reply:
Just realized while playing around that we *need* the certs secrets to be available as long as Pods using them still exist, because ECK needs to be able to make https requests to the cluster, even though a rolling upgrade to disable TLS might be scheduled.
So if we remove certs secrets it must be after the rolling upgrade from non-TLS to TLS is over.
@charith-elastic how about we scope this out of this PR, and I create a dedicated low priority issue?

Reply:
Of course, I was simply curious why Elasticsearch needed a special case. Perfectly happy with deferring it to another issue.

Reply:
Created https://github.com/elastic/cloud-on-k8s/issues/2766.